src/event.rs

```diff
@@ -16,138 +16,33 @@
-                    destination  : match orig_message {
-                                       &Some(ref signed_message) => {
-                                           let routing_message = try!(signed_message.get_routing_message());
-                                           routing_message.reply_destination()
-                                       }
-                                       &None => DestinationAddress::Direct(source_group.clone()),
-                                   },
-                    //destination  : match orig_message.get_routing_message() {
-                    //                    Ok(routing_message) => routing_message.reply_destination(),
-                    //                    Err(_) => DestinationAddress::Direct(source_group.clone()),
-                    //               },
-                    source       : SourceAddress::Direct(destination_group.clone()),
-                    orig_message : None,
-                    message_type : reply_data,
-                    message_id   : message_id.clone(),
-                    authority    : our_authority.clone()
-                })
-            },
-            _ => Err(RoutingError::Response(ResponseError::InvalidRequest))
-        }
-    }
-
-    pub fn get_orig_message(&self) -> Option<SignedMessage> {
-        match self {
-            &Event::PutDataRequest(ref orig_message, _, _, _, _, _, _)
-                => orig_message.clone(),
-            &Event::PutDataResponse(ref orig_message, _, _, _, _, _, _)
-                => orig_message.clone(),
-            &Event::GetDataResponse(ref orig_message, _, _, _, _, _, _)
-                => orig_message.clone()
-        }
-    }
-
-    pub fn get_data(&self) -> Result<Data, RoutingError> {
-        match self {
-            &Event::PutDataRequest(_, ref data, _, _, _, _, _) => Ok(data.clone()),
-            _ => Err(RoutingError::Response(ResponseError::InvalidRequest))
-        }
-    }
-
-    pub fn get_response(&self) -> Result<ErrorReturn, RoutingError> {
-        match self {
-            &Event::PutDataResponse(_, ref response, _, _, _, _, _) => Ok(response.clone()),
-            _ => Err(RoutingError::Response(ResponseError::InvalidRequest))
-        }
-    }
-
-    pub fn get_data_response(&self) -> Result<GetDataResponse, RoutingError> {
-        match self {
-            &Event::GetDataResponse(_, ref response, _, _, _, _, _) => Ok(response.clone()),
-            _ => Err(RoutingError::Response(ResponseError::InvalidRequest))
-        }
-    }
-}
-
-impl Source<NameType> for Event {
-    fn get_source(&self) -> NameType {
-        match self {
-            &Event::PutDataRequest(_, _, source_group, _, _, _, _) => source_group,
-            &Event::PutDataResponse(_, _, source_group, _, _, _, _) => source_group,
-            &Event::GetDataResponse(_, _, source_group, _, _, _, _) => source_group,
-        }
-    }
+    MessageSecured(RoutingMessage, Authority),
```

Comment:
I think possibly messages also `To` Clients / ManagedNodes (I am thinking of PmidNodes) need to be returned where they cannot take an action the network requests. I agree usually a group message is consensus checked and does not perhaps need to original signed message though. Interesting observation! 

Possibly if a group gives us back a error we can consider we asked for it, so security wise that is fine, the other side of the coin is getting the actual request back as the network cannot fix it, meaning we need the original message to be able to resolve the error.